Release notes and migration guide ready to merge (#2370)
tinymce/tinymce-docs: modules/ROOT/pages/6.0-release-notes-plugins.adoc

= TinyMCE 6.0 Plugin changes
:navtitle: TinyMCE 6.0 Plugin changes
:description: TinyMCE 6.0 Plugin changes
:keywords: releasenotes, plugin

This section documents the improvements, changes, and removals made to {productname} 6.0 Plugins.

It also documents the the plugin loading change introduced in {productname} 6.0.

* xref:plugins[Plugins]
* xref:plugin-loading-format-change[Plugin loading format change]
* xref:new-and-improved-plugins][New and improved Plugins]
* xref:changed-apis[Changed APIs]
* xref:removed-or-deprecated-plugins[Removed or deprecated Plugins]

// tag::plugins[]
[[plugins]]
== Plugins

[[plugin-loading-format-change]]
=== Plugin loading format change

{productname} 6.0 allows plugin loading in any of three formats:

. An array of strings:
+
`plugins: [ 'advlist', 'list', 'image', 'help', 'wordcount' ]`
+
. A space-separated string:
+
`plugins: 'advlist list image help wordcount'`
+
. A comma-separated string:
+
`plugins: 'advlist,list,image,help,wordcount'`

> Previously, two other formats were allowed: comma-separated strings with spaces and arrays of space-separated strings.

Reviewer: ```suggestion
Previously, one other format was allowed: arrays of space-separated strings.
```

Sentence following this also needs editing apparently

Neither of these formats are allowed in {productname} 6.0.

[[new-and-improved-plugins]]
=== New and improved plugins

[[imagetools]]
==== EditImage

_EditImage_, a https://tiny.cloud/tinymce/features/#productivity[Premium plugin], replaces the now-deprecated and removed _ImageTools_ plugin. The `imagetools` plugin is no longer part of the Core open source editor.

> An `editimage` options were also renamed: `editimage_proxy` is now `editimage_proxy_service_url`, making it easier to configure when accessing https://tiny.cloud/docs/enterprise/server/[Tiny-specific services].

Reviewer 2: This sentence doesn't seem to flow and is quite confusing to me.

`editimage` also offers a new export plugin option: `export_image_proxy_service_url`.

NOTE: `editimage_proxy` (the renamed equivalent to the previous `imagetools_proxy`) is still available as an alternative configuration option for setting up image proxy functionality without using Tiny services.

[[emoticons]]
==== Emoticons

The `emoticon` plugin now uses the more accurate word, _Emoji_, in its tool bar, menu item and dialog box.

In previous versions, the word presented in these parts of the plugin’s user-interface was _Emoticon_.

NOTE: The plugin, itself, is still called Emoticons. And the plugin description and the plugin’s configuration and Help dialogs still reflect this.

[[table-of-contents]]

> ==== Table of contents

Reviewer 3: ```suggestion
==== Table of Contents
```
This is it's official name 🤷

> There are no end-user facing changes between the previous `toc` plugin and the `tableofcontents` plugin. However, the `tableofcontents` plugin now uses https://terser.org[terser] to improve performance.

Reviewer 3: ```suggestion
There are no end-user facing changes between the previous `toc` plugin and the `tableofcontents` plugin.
```
This statement is incorrect and has nothing to do with performance, so we should just remove it.

NOTE: the `tableofcontents` plugin is no longer part of the Core open source editor. Table of Contents is now a https://tiny.cloud/tinymce/features/#productivity[Premium plugin].

> [[]]

Reviewer 3: We should either add an id here or just remove this.

> ==== Hindi, Malay, and Vietnamese translations added to four Premium Plugins

Reviewer 3: This also isn't quite correct... it was added for the following:
- Accessibility Checker
- Advanced Table
- Comments
- Enhanced Image Editing
- Export
- PowerPaste
- Real-time Collaboration (RTC)
- Spell Checker Pro
- Table of Contents

The following four Premium Plugins now present in Hindi, Malay, and Vietnamese:

* Accessibility Checker

> * AdvTable

Reviewer 3: ```suggestion
* Advanced Table
```

* Comments

> * Spellchecker

Reviewer 3: ```suggestion
* Spell Checker Pro
```

==== Noneditable

The `noneditable` plugin is no longer a plugin. It is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `noneditable` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

[[paste]]
==== Paste

The `paste` plugin’s functionality is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `paste` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

The `paste_data_images` option now defaults to `true`. When `paste` was a plugin, this option was, by default, set to `false`, which prevented images being pasted from the local machine.

The `mceInsertClipboardContent` argument, `content`, has been renamed. It is now `html`.

The new name is a more accurate reflection of what sort of data the argument passes.

> 
> NOTE: `PastePreProcess` also takes an argument, `content`. This argument, for this command, has not changed.

Reviewer 3: ```suggestion
```
This isn't correct, as it's an event so it can't take arguments. I think we can just remove this line.

[[print]]
==== Print

`print` functionality is no longer a plugin. It is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `print` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

[[table]]
==== Table

The `table` plugin’s functionality is, as of {productname} 6.0, entirely focussed on the user interface (UI) for creating and editing tables.

The code that generates tables is now a Core part of {productname} 6.0.

This, now core feature of {productname} 6.0, also includes the following fixes, changes, and improvements.

> * If the selected cells in the `tfoot` section of a table were header elements (ie were `th` elements), pressing the `table` row header toolbar button (which invokes the `mceTableRowType` command) returned the row as a header row.

Reviewer 3: Should this be a list? It seems like it should just be a paragraph or something:
![image](https://user-images.githubusercontent.com/1575550/161907633-9bbdc854-ff48-4293-9292-03f2b8de5945.png)

> * If the selected cells in the `tfoot` section of a table were header elements (ie were `th` elements), pressing the `table` row header toolbar button (which invokes the `mceTableRowType` command) returned the row as a header row.
> 
> This was incorrect: the section takes precedence and cells in the `tfoot` section of a table should be declared as a footer row.
> 
> In {productname} 6.0, selecting cells in the `tfoot` section of a table and pressing the row header toolbar button returns the row as a footer row, as expected.
> 

Reviewer: This reads weirdly to me, the way its rendering. I think it's all meant to be one bullet point?
```suggestion
* If the selected cells in the `tfoot` section of a table were header elements (ie were `th` elements), pressing the `table` row header toolbar button (which invokes the `mceTableRowType` command) returned the row as a header row.
+
This was incorrect: the section takes precedence and cells in the `tfoot` section of a table should be declared as a footer row.
+
In {productname} 6.0, selecting cells in the `tfoot` section of a table and pressing the row header toolbar button returns the row as a footer row, as expected.

```

* `mceInsertTableDialog` is the new command for invoking an _Insert Table_ dialog box.
+
The previous command, `mceInsertTable`, no longer works to invoke an _Insert Table_ dialog box: use the `mceInsertTableDialog` command instead.
+
The `mceInsertTable` command is still the command for inserting a table with specified properties directly and without asking for user-input.
+
* The default values for the `table_style_by_css` and the `table_use_colgroups` options have both been changed.
+
Previously, these options defaulted to `false`. As of {productname} 6.0, both options now default to `true`.
+
This is in line with modern web practice.
+
* the `table_responsive_width` option was previously deprecated and superseded by `table_sizing_mode`.
+
In {productname} 6.0, the `table_responsive_width` option has been removed.

> //[[fixed-plugins]]
> //=== Fixed plugins
> //
> //[id-added-plugins]]
> //=== Added plugins
> 

Reviewer 3: ```suggestion
```
Since we have nothing to add here, we might as well remove it?

[[removed-or-deprecated-plugins]]
=== Removed or deprecated plugins

[[bbcode]]
==== BBCode

> As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the `bbcode` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.

Reviewer 3: ```suggestion
As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the `bbcode` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.
```
As this is a different set of docs we should open in a new tab/window. @metricjs also suggested mentioning it's `TinyMCE 5` to make it clearer it navigates away.

[[color-picker]]
==== Color Picker

Color Picker’s functionality has been a built in part of {productname} since version 5.

> Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the empty `colorpicker` plugin has been removed.

Reviewer 3: ```suggestion
Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the empty `colorpicker` plugin has been removed.
```

[[context-menu]]
==== Context Menu

Context Menu’s functionality has been a built in part of {productname} since version 5.

> Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the empty `contextmenu` plugin has been removed.

Reviewer 3: ```suggestion
Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the empty `contextmenu` plugin has been removed.
```

[[full-page]]
==== Full Page

> As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the `fullpage` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.

Reviewer 3: ```suggestion
As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the `fullpage` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.
```

[[hr]]
==== HR

The `hr` (horizontal rule) functionality is no longer a plugin. It is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `hr` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

[[changed-plugins-imagetools]]
==== ImageTools

The `imagetools` plugin is no longer part of the Free {productname} editor. _Image Tools_ is now _Edit Image_, a https://tiny.cloud/tinymce/features/#productivity[Premium plugin].

[[legacy-output]]
==== Legacy Output

> As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the `legacyoutput` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.

Reviewer 3: ```suggestion
As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the `legacyoutput` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.
```

[[changed-plugins-paste]]
==== Paste

The `paste` functionality is no longer a plugin. It is now a Core part of {productname} 6.0.

> Because the functionality was unmaintained, `paste` no longer supports input from Microsoft Word.

Reviewer: Though this was kind of the reason, I'm not sure we should really say "because it was unmaintained" in the docs?

Reviewer: If anything, maybe "unmaintainable" might be better - both on implications and accuracy

There is an open https://github.com/tinymce/tinymce/discussions/7487[request for maintainers] should someone in the community be interested in taking over maintainance of this particular functionality as a separate plugin.

If a community-maintained version of the `paste` plugin becomes available, we will link to it from the xref:migration-from-5x.adoc#plugins-paste[_Migration Guide_].

> NOTE: the Premium plugin, https://tiny.cloud/docs/plugins/premium/powerpaste/[`PowerPaste`], is available. It provides the capability to accept data from Microsoft Word and Microsoft Excel, and clean-up the received data before pasting it into place.

Reviewer: ```suggestion
NOTE: The Premium plugin, https://tiny.cloud/docs/plugins/premium/powerpaste/[`PowerPaste`], is available. It provides the capability to accept data from Microsoft Word and Microsoft Excel, and clean-up the received data before pasting it into place.
```

Reviewer 3: ```suggestion
NOTE: The Premium plugin, xref:introduction-to-powerpaste.adoc[`PowerPaste`], is available. It provides the capability to accept data from Microsoft Word and Microsoft Excel, and clean-up the received data before pasting it into place.
```

We should be linking to the current docs, not TinyMCE 5 docs.

[[changed-plugins-print]]
==== Print

`print` functionality is no longer a plugin. It is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `print` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

[[spellchecker]]
==== Spellchecker

> As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the `spellchecker` plugin has been removed. It is no longer part of {productname} 6.0.

Reviewer 3: ```suggestion
As per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the `spellchecker` plugin has been removed. It is no longer part of {productname} 6.0.
```

> The https://tiny.cloud/tinymce/features/#productivity[Premium plugin], https://tiny.cloud/docs/plugins/premium/tinymcespellchecker/[Spellchecker Pro], which offers equivalent functionality and more, is available.

Reviewer 3: ```suggestion
The https://tiny.cloud/tinymce/features/#productivity[Premium plugin], xref:introduction-to-tiny-spellchecker.adoc[Spell Checker Pro], which offers equivalent functionality and more, is available.
```

[[tabfocus]]
==== TabFocus

The `tabfocus` plugin has been removed. It is no longer part of {productname} 6.0. And it is not otherwise available.

The `tabfocus_elements` option, which relied on the `tabfocus` plugin, no longer functions.

As well, the `tab_focus` option was previously deprecated by `tabfocus_elements`. With this release, the `tab_focus` option has been removed completely.

As per standard web practice, the `tabindex` attribute should be used instead of the `tabfocus` plugin or any of the `tabfocus` plugin’s options.

As per that standard web practice, the `tabindex` attribute is copied from the target element to the iframe.

[[changed-plugins-table-of-contents]]
==== Table of contents

The `toc` plugin is no longer part of the Core open source editor. _Table of Contents_ is now a https://tiny.cloud/tinymce/features/#productivity[Premium plugin].

> NOTE: although there are no end-user facing changes, as part of the general architectural change, the `tableofcontents` Premium plugin has modernised its output. It uses the ES2020 library and targets ES2018. It also uses https://terser.org[terser] to improve performance.

Reviewer: ```suggestion
NOTE: Although there are no end-user facing changes, as part of the general architectural change, the `tableofcontents` Premium plugin has modernised its output. It uses the ES2020 library and targets ES2018. It also uses https://terser.org[terser] to improve performance.
```

Reviewer 3: Similar to my previous comment, the terser change really doesn't have anything to do with performance. It's a minifier so the only perf gain you can get from that is loading speed and we're not going to get anything noticeable from swapping from UglifyJS to Terser. We also shouldn't mention anything about ES2020 library/targets as that's an internal build thing and would likely only confuse customers.
```suggestion
NOTE: Although there are no end-user facing changes, as part of the general architectural change, the `tableofcontents` Premium plugin has modernised its output.
```

[[text-color]]
==== Text color

Text color’s functionality has been a built in part of {productname} since version 5.

> Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[announcement], the empty `textcolor` plugin has now been removed.

Reviewer 3: ```suggestion
Consequently, and as per the previous https://www.tiny.cloud/docs/release-notes/6.0-upcoming-changes/[{productname} 5 announcement, window="_blank"], the empty `textcolor` plugin has now been removed.
```

[[text-pattern]]
==== TextPattern

The `textpattern` functionality is no longer a plugin. It is now a Core part of {productname} 6.0.

In {productname} 6.0, there is no setting or specifying this functionality in the `plugins` option.

All references to `textpattern` should be removed from your plugin configuration. Leaving references to removed plugins in a plugin configuration will cause console errors documenting the plugin failing to load.

> Also, the API that called `textpattern` has been removed.

Reviewer 2: This is another place where we could mention that the API has been replaced with calls to `editor.options.get('text_patterns')` and `editor.options.set('text_patterns', newPatterns)`.

Reviewer 3: The text pattern API wasn't actually ever documented, so we possibly shouldn't actually even mention it here.

Reviewer: That is a good point

Reviewer: ```suggestion
```
